source/lac/cuda_vector.cu

```diff
@@ -0,0 +1,1006 @@
+#include <deal.II/lac/read_write_vector.h>
+#include <deal.II/base/exceptions.h>
+#include <cmath>
+
+#ifdef DEAL_II_WITH_CUDA
+
+DEAL_II_NAMESPACE_OPEN
+
+#define BKSIZE_ELEMWISE_OP 512
+#define CHUNKSIZE_ELEMWISE_OP 8
+#define VR_BKSIZE 512
+#define VR_CHUNK_SIZE 8
+
+namespace LinearAlgebra
+{
+  namespace CUDAWrappers
+  {
+    namespace internal
+    {
+      template <typename Number>
+      __global__ void vec_scale(Number                                   *val,
+                                const Number                              a,
+                                const typename Vector<Number>::size_type  N)
+      {
+        const typename Vector<Number>::size_type idx_base = threadIdx.x +
+                                                            blockIdx.x *
+                                                            (blockDim.x*CHUNKSIZE_ELEMWISE_OP);
+        for (unsigned int i=0; i<CHUNKSIZE_ELEMWISE_OP; ++i)
+          {
+            const typename Vector<Number>::size_type idx = idx_base +
+                                                           i*BKSIZE_ELEMWISE_OP;
+            if (idx<N)
+              val[idx] *= a;
+          }
+      }
+
+
+
+      struct Binop_Addition
+      {
+        template <typename Number>
+        __device__ static inline Number operation(const Number a,
+                                                  const Number b)
+        {
+          return a+b;
+        }
+      };
+
+
+
+      struct Binop_Subtraction
+      {
+        template <typename Number>
+        __device__ static inline Number operation(const Number a,
+                                                  const Number b)
+        {
+          return a-b;
+        }
+      };
+
+
+
+      template <typename Number, typename Binop>
+      __global__ void vector_bin_op(Number                                   *v1,
+                                    Number                                   *v2,
+                                    const typename Vector<Number>::size_type  N)
+      {
+        const typename Vector<Number>::size_type idx_base = threadIdx.x +
+                                                            blockIdx.x *
+                                                            (blockDim.x*CHUNKSIZE_ELEMWISE_OP);
+        for (unsigned int i=0; i<CHUNKSIZE_ELEMWISE_OP; ++i)
+          {
+            const typename Vector<Number>::size_type idx = idx_base +
+                                                           i*BKSIZE_ELEMWISE_OP;
+            if (idx<N)
+              v1[idx] = Binop::operation(v1[idx],v2[idx]);
+          }
+      }
+
+
+      template <typename Number>
+      struct L1Norm
+      {
+        __device__ static Number reduction_op(const Number a, const Number b)
+        {
+          return std::abs(a) + std::abs(b);
+        }
+
+        __device__ static void atomic_op(Number *dst, const Number a)
+        {
+          *dst = std::abs(*dst) + std::abs(a);
```

Comment:
By the way, can't you conclude by induction that you don't need the first of the two `abs` calls in
```
*dst = std::abs(*dst) + std::abs(a);
```
?